In [ ]:
from google.colab import files
from IPython.display import Image
uploaded =files.upload()

Saving ASML.png to ASML.png


# Introduction

When it comes to choosing a new phone, battery life and camera quality are important factors to consider. A phone with a long battery life can save you from the hassle of constantly searching for an outlet to charge your phone. Additionally, a phone with high-quality pixels can capture stunning photos and videos that can be cherished for years to come. While new phones may be released every six months, taking the time to research and choose a phone with the features that matter most to you can result in a satisfying and long-lasting purchase.

The data set contains 21 columns


# **Importing libraries**

In [ ]:
# impot libraries
import pandas as pd
import numpy as np
import sklearn.model_selection
from sklearn.preprocessing import StandardScaler
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from sklearn import metrics
from sklearn.metrics import accuracy_score

In [ ]:
df = pd.read_csv('/content/train.csv')
df.head(5)

battery_power  blue  clock_speed  dual_sim  fc  four_g  int_memory  m_dep  \
0            842     0          2.2         0   1       0           7    0.6   
1           1021     1          0.5         1   0       1          53    0.7   
2            563     1          0.5         1   2       1          41    0.9   
3            615     1          2.5         0   0       0          10    0.8   
4           1821     1          1.2         0  13       1          44    0.6   

   mobile_wt  n_cores  ...  px_height  px_width   ram  sc_h  sc_w  talk_time  \
0        188        2  ...         20       756  2549     9     7         19   
1        136        3  ...        905      1988  2631    17     3          7   
2        145        5  ...       1263      1716  2603    11     2          9   
3        131        6  ...       1216      1786  2769    16     8         11   
4        141        2  ...       1208      1212  1411     8     2         15   

   three_g  touch_screen  wifi  price_range  
0        0             0     1            1  
1        1             1     0            2  
2        1             1     0            2  
3        1             0     0            2  
4        1             1     0            1  

[5 rows x 21 columns]

# General overview of the data

In [ ]:
df.describe()

battery_power       blue  clock_speed     dual_sim           fc  \
count    2000.000000  2000.0000  2000.000000  2000.000000  2000.000000   
mean     1238.518500     0.4950     1.522250     0.509500     4.309500   
std       439.418206     0.5001     0.816004     0.500035     4.341444   
min       501.000000     0.0000     0.500000     0.000000     0.000000   
25%       851.750000     0.0000     0.700000     0.000000     1.000000   
50%      1226.000000     0.0000     1.500000     1.000000     3.000000   
75%      1615.250000     1.0000     2.200000     1.000000     7.000000   
max      1998.000000     1.0000     3.000000     1.000000    19.000000   

            four_g   int_memory        m_dep    mobile_wt      n_cores  ...  \
count  2000.000000  2000.000000  2000.000000  2000.000000  2000.000000  ...   
mean      0.521500    32.046500     0.501750   140.249000     4.520500  ...   
std       0.499662    18.145715     0.288416    35.399655     2.287837  ...   
min       0.000000     2.000000     0.100000    80.000000     1.000000  ...   
25%       0.000000    16.000000     0.200000   109.000000     3.000000  ...   
50%       1.000000    32.000000     0.500000   141.000000     4.000000  ...   
75%       1.000000    48.000000     0.800000   170.000000     7.000000  ...   
max       1.000000    64.000000     1.000000   200.000000     8.000000  ...   

         px_height     px_width          ram         sc_h         sc_w  \
count  2000.000000  2000.000000  2000.000000  2000.000000  2000.000000   
mean    645.108000  1251.515500  2124.213000    12.306500     5.767000   
std     443.780811   432.199447  1084.732044     4.213245     4.356398   
min       0.000000   500.000000   256.000000     5.000000     0.000000   
25%     282.750000   874.750000  1207.500000     9.000000     2.000000   
50%     564.000000  1247.000000  2146.500000    12.000000     5.000000   
75%     947.250000  1633.000000  3064.500000    16.000000     9.000000   
max    1960.000000  1998.000000  3998.000000    19.000000    18.000000   

         talk_time      three_g  touch_screen         wifi  price_range  
count  2000.000000  2000.000000   2000.000000  2000.000000  2000.000000  
mean     11.011000     0.761500      0.503000     0.507000     1.500000  
std       5.463955     0.426273      0.500116     0.500076     1.118314  
min       2.000000     0.000000      0.000000     0.000000     0.000000  
25%       6.000000     1.000000      0.000000     0.000000     0.750000  
50%      11.000000     1.000000      1.000000     1.000000     1.500000  
75%      16.000000     1.000000      1.000000     1.000000     2.250000  
max      20.000000     1.000000      1.000000     1.000000     3.000000  

[8 rows x 21 columns]

In [ ]:
df['price_range'].value_counts()

1    500
2    500
3    500
0    500
Name: price_range, dtype: int64

Here i just splited the data horizontaly and  for the test_size.Taking 80% as a training set and 20% as test data.

In [ ]:
df_train, df_test = sklearn.model_selection.train_test_split(df, test_size = 0.2)

# Printing the size of train and test sets
print("df_train size:", df_train.shape)
print("df_test size:", df_test.shape)

df_train size: (1600, 21)
df_test size: (400, 21)


# **Data preprocessing**

We just identify our x and y as train and test variables


In [ ]:

#separating the features annd the target label

X_train = df_train.drop(['price_range'], axis = 1)
y_train =  df_train['price_range']

X_test = df_test.drop(['price_range'], axis = 1)
y_test = df_test['price_range']
#The shape of each data
print('x_train size:', X_train.shape)
print('x_test size:', X_test.shape)
print('y_train size:', y_train.shape)
print('y_test size:', y_test.shape)


x_train size: (1600, 20)
x_test size: (400, 20)
y_train size: (1600,)
y_test size: (400,)


Our all ocategorical data are already in numerical form, then  no need to use encoding techniques such as one-hot encoding or label encoding.



scaling data is necessary because features have different scale

In [ ]:

#scaling our data to make all the values will be  changed to the same  range
scaler = sklearn.preprocessing.StandardScaler(with_mean=False, with_std = True)
scaler.fit(X_train)

#scaling the training features
X_train = scaler.transform(X_train)

#scaling the training x_test
X_test = scaler.transform(X_test)

print('x_train is:', X_train)
print('x_test is :', X_test)



x_train is: [[3.43743735 2.00026411 3.05235233 ... 0.         0.         0.        ]
 [3.85175099 2.00026411 2.56397595 ... 2.36950204 0.         0.        ]
 [3.48296632 0.         2.44188186 ... 0.         0.         2.00068942]
 ...
 [2.41075904 0.         0.61047047 ... 0.         2.00010001 2.00068942]
 [2.11709717 0.         2.56397595 ... 0.         2.00010001 0.        ]
 [2.64750969 0.         3.41863461 ... 0.         0.         2.00068942]]
x_test is : [[3.86768613 2.00026411 0.61047047 ... 2.36950204 2.00010001 0.        ]
 [2.82279623 0.         1.95350549 ... 0.         0.         2.00068942]
 [3.5945123  2.00026411 0.61047047 ... 0.         0.         2.00068942]
 ...
 [2.17856129 2.00026411 3.05235233 ... 2.36950204 0.         2.00068942]
 [2.52458147 2.00026411 3.41863461 ... 2.36950204 0.         0.        ]
 [1.55253793 2.00026411 0.61047047 ... 0.         2.00010001 2.00068942]]


here we can see in this matrices that all of our data is scaled and have the same range between 0 and 4

Here ae are using Decision Tree Classifier because we predict continuous
numerical values, as a hyperparameter i set  max depth to 9 that number of levels that the tree can grow,and we set for criterion a mean squared error  that ,seeks to minimize the average squared difference between anticipated and actual values.

We use gridsearch to find the optimal combination of hyperparameters for the models

In [ ]:
# Choosing our hyperparameters for DecisionTreeRegressor
dtree =  sklearn.tree.DecisionTreeClassifier()
parameters_dt = {'criterion':['gini', 'entropy', 'log_loss'],'max_depth':[1,3,5], 'min_samples_split':[2,6,8]}


# defining parameter range
dtree_grid = sklearn.model_selection.GridSearchCV(dt, parameters_dt, scoring='accuracy' ,cv=10, verbose=1)

# fitting the model for grid search
dtree_grid.fit(X_train,y_train)

#printing the results
print(f'the best score of the decision tree is:{dtree_grid.best_score_}')
print(f'the best parameters are : {dtree_grid.best_params_}')

Fitting 10 folds for each of 27 candidates, totalling 270 fits
the best score of the decision tree is:0.82125
the best parameters are : {'criterion': 'gini', 'max_depth': 5, 'min_samples_split': 2}


According to the results, a decision tree model was trained and tested using cross-validation,Here fitting the model to the training dataset and evaluating its performance on 10 different folds of the data.
Because there were 27 different hyperparameters combinations being evaluated, this technique was raning 270 times.The score represents a performance metric using accuracy.

if we set cross_validation to lower or upper values we will ge tlow accuracy .

Here we are using Support Vector Machines SVC as a model to then we set  some rndom hyperparameters and use gridsearch to fid the best hyperparameters.

In [ ]:
# Choosing our hyperparameters for s.v.m
svm = SVC()
parameters_svm = {'C':[2,3,5], 'kernel':('linear', 'poly', 'rbf', 'sigmoid'), 'random_state':[5,10,15,20]}

# defining parameter range
svm_grid = sklearn.model_selection.GridSearchCV(svm, parameters_svm, scoring='accuracy' ,cv=10, verbose=1)

# fitting the model for grid search
svm_grid.fit(X_train,y_train)

#ptinting the best score and best parameters
print(f'the best score of the decision tree is:{svm_grid.best_score_}')
print(f'the best parameters is : {svm_grid.best_params_}')

Fitting 10 folds for each of 48 candidates, totalling 480 fits
the best score of the decision tree is:0.9668750000000002
the best parameters is : {'C': 3, 'kernel': 'linear', 'random_state': 5}


A grid search was used to discover the optimum hyperparameters for a support vector machine (SVM) model, according to the result . The grid search includes training and testing the model with 10-fold cross-validation on 640 distinct hyperparameter combinations.The score represents the performance  using accuracy metric

# Prediction

In [ ]:
# Making final predicitions
import sklearn.metrics
y_predicted = svm_grid.predict(X_test)
# Calculate and save the accuracy score
accuracy_score = sklearn.metrics.accuracy_score(y_predicted, y_test)

# Print the accuracy score
print("The  accuracy score for svm model is {}%".format(accuracy_score))

The  accuracy score for svm model is 0.9625%


# Conclusion

Based on the scores, I conclude that the Support Vector Machine (SVM) model is the best model for my data compared to the Decision Tree model. By using grid search, we were able to define the best hyperparameters for our SVM model, resulting in improved accuracy during training. This improved accuracy can be observed in our final model performance wich is 0.96 close to our result in training This means that our model training is flawless and that it learnt effectively from the training dataset.

We can now predict the best price range for a given phone by inputting its relevant features. By considering factors such as battery power, camera etc.




In [ ]:
!jupyter nbconvert --to html Machin__learning__mini_pro.ipynb

[NbConvertApp] Converting notebook Machin__learning__mini_pro.ipynb to html
[NbConvertApp] Writing 719607 bytes to Machin__learning__mini_pro.html


**Reference**:

(ABHISHEK SHARMA, 2017)
https://www.kaggle.com/datasets/iabhishekofficial/mobile-price-classification Accsess (Mai 11,2023)
